File: E-Commerce Inventory.cpp
#include <iostream>
#include <fstream>
#include <unordered_map>
#include <sstream>
#include <iomanip>

struct Product {
    std::string id;
    std::string name;
    double price;
    int quantity;
};

class InventoryManager {
private:
    std::unordered_map<std::string, Product> inventory;

public:
    void loadFromFile(const std::string& filename) {
        std::ifstream file(filename);
        std::string line;

        if (!file) {
            std::cerr << "Warning: Could not open file. Starting with empty inventory.\n";
            return;
        }

        while (getline(file, line)) {
            std::stringstream ss(line);
            Product p;
            std::string priceStr, quantityStr;

            getline(ss, p.id, ',');
            getline(ss, p.name, ',');
            getline(ss, priceStr, ',');
            getline(ss, quantityStr, ',');

            p.price = std::stod(priceStr);
            p.quantity = std::stoi(quantityStr);

            inventory[p.id] = p;
        }

        file.close();
    }

    void saveToFile(const std::string& filename) {
        std::ofstream file(filename);
        for (const auto& [id, product] : inventory) {
            file << product.id << "," << product.name << ","
                 << product.price << "," << product.quantity << "\n";
        }
        file.close();
    }

    void addProduct() {
        Product p;
        std::cout << "Enter Product ID: ";
        std::cin >> p.id;

        if (inventory.find(p.id) != inventory.end()) {
            std::cout << "Product with this ID already exists.\n";
            return;
        }

        std::cin.ignore();
        std::cout << "Enter Product Name: ";
        getline(std::cin, p.name);
        std::cout << "Enter Price: ";
        std::cin >> p.price;
        std::cout << "Enter Quantity: ";
        std::cin >> p.quantity;

        inventory[p.id] = p;
        std::cout << "Product added successfully.\n";
    }

    void updateProduct() {
        std::string id;
        std::cout << "Enter Product ID to update: ";
        std::cin >> id;

        auto it = inventory.find(id);
        if (it == inventory.end()) {
            std::cout << "Product not found.\n";
            return;
        }

        Product& p = it->second;
        std::cout << "Enter new name [" << p.name << "]: ";
        std::cin.ignore();
        getline(std::cin, p.name);
        std::cout << "Enter new price [" << p.price << "]: ";
        std::cin >> p.price;
        std::cout << "Enter new quantity [" << p.quantity << "]: ";
        std::cin >> p.quantity;

        std::cout << "Product updated successfully.\n";
    }

    void deleteProduct() {
        std::string id;
        std::cout << "Enter Product ID to delete: ";
        std::cin >> id;

        if (inventory.erase(id)) {
            std::cout << "Product deleted.\n";
        } else {
            std::cout << "Product not found.\n";
        }
    }

    void searchProduct() {
        std::string id;
        std::cout << "Enter Product ID to search: ";
        std::cin >> id;

        auto it = inventory.find(id);
        if (it != inventory.end()) {
            const Product& p = it->second;
            std::cout << "ID: " << p.id << "\n"
                      << "Name: " << p.name << "\n"
                      << "Price: $" << p.price << "\n"
                      << "Quantity: " << p.quantity << "\n";
        } else {
            std::cout << "Product not found.\n";
        }
    }

    void listProducts() {
        std::cout << std::left << std::setw(10) << "ID" 
                  << std::setw(20) << "Name"
                  << std::setw(10) << "Price"
                  << std::setw(10) << "Quantity" << "\n";

        std::cout << "-----------------------------------------------------\n";
        for (const auto& [id, p] : inventory) {
            std::cout << std::left << std::setw(10) << p.id
                      << std::setw(20) << p.name
                      << std::setw(10) << p.price
                      << std::setw(10) << p.quantity << "\n";
        }
    }
};

void displayMenu() {
    std::cout << "\n--- Inventory Management ---\n";
    std::cout << "1. Add Product\n";
    std::cout << "2. Update Product\n";
    std::cout << "3. Delete Product\n";
    std::cout << "4. Search Product\n";
    std::cout << "5. List All Products\n";
    std::cout << "6. Save & Exit\n";
    std::cout << "Choose an option: ";
}

int main() {
    InventoryManager manager;
    manager.loadFromFile("inventory.csv");

    int choice;
    do {
        displayMenu();
        std::cin >> choice;
        switch (choice) {
            case 1: manager.addProduct(); break;
            case 2: manager.updateProduct(); break;
            case 3: manager.deleteProduct(); break;
            case 4: manager.searchProduct(); break;
            case 5: manager.listProducts(); break;
            case 6:
                manager.saveToFile("inventory.csv");
                std::cout << "Data saved. Exiting...\n";
                break;
            default: std::cout << "Invalid option.\n"; break;
        }
    } while (choice != 6);

    return 0;
}

File: E-Commerce-inventory.java
import java.io.*;
import java.util.*;

class Product {
    String id;
    String name;
    double price;
    int quantity;

    Product(String id, String name, double price, int quantity) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.quantity = quantity;
    }

    @Override
    public String toString() {
        return String.format("%-10s %-20s $%-10.2f %-10d", id, name, price, quantity);
    }
}

public class InventoryManager {
    private final Map<String, Product> inventory = new HashMap<>();
    private final String fileName = "inventory.csv";
    private final Scanner scanner = new Scanner(System.in);

    public void loadFromFile() {
        File file = new File(fileName);
        if (!file.exists()) {
            System.out.println("Warning: File not found. Starting with an empty inventory.");
            return;
        }

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", -1);
                if (parts.length == 4) {
                    String id = parts[0];
                    String name = parts[1];
                    double price = Double.parseDouble(parts[2]);
                    int quantity = Integer.parseInt(parts[3]);
                    inventory.put(id, new Product(id, name, price, quantity));
                }
            }
        } catch (IOException e) {
            System.out.println("Error reading file.");
        }
    }

    public void saveToFile() {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(fileName))) {
            for (Product p : inventory.values()) {
                bw.write(String.format("%s,%s,%f,%d%n", p.id, p.name, p.price, p.quantity));
            }
        } catch (IOException e) {
            System.out.println("Error writing file.");
        }
    }

    public void addProduct() {
        System.out.print("Enter Product ID: ");
        String id = scanner.nextLine();

        if (inventory.containsKey(id)) {
            System.out.println("Product with this ID already exists.");
            return;
        }

        System.out.print("Enter Product Name: ");
        String name = scanner.nextLine();

        System.out.print("Enter Price: ");
        double price = scanner.nextDouble();

        System.out.print("Enter Quantity: ");
        int quantity = scanner.nextInt();
        scanner.nextLine(); // consume newline

        inventory.put(id, new Product(id, name, price, quantity));
        System.out.println("Product added successfully.");
    }

    public void updateProduct() {
        System.out.print("Enter Product ID to update: ");
        String id = scanner.nextLine();

        Product p = inventory.get(id);
        if (p == null) {
            System.out.println("Product not found.");
            return;
        }

        System.out.print("Enter new name [" + p.name + "]: ");
        String name = scanner.nextLine();
        if (!name.isEmpty()) p.name = name;

        System.out.print("Enter new price [" + p.price + "]: ");
        String priceStr = scanner.nextLine();
        if (!priceStr.isEmpty()) p.price = Double.parseDouble(priceStr);

        System.out.print("Enter new quantity [" + p.quantity + "]: ");
        String qtyStr = scanner.nextLine();
        if (!qtyStr.isEmpty()) p.quantity = Integer.parseInt(qtyStr);

        System.out.println("Product updated successfully.");
    }

    public void deleteProduct() {
        System.out.print("Enter Product ID to delete: ");
        String id = scanner.nextLine();

        if (inventory.remove(id) != null) {
            System.out.println("Product deleted.");
        } else {
            System.out.println("Product not found.");
        }
    }

    public void searchProduct() {
        System.out.print("Enter Product ID to search: ");
        String id = scanner.nextLine();

        Product p = inventory.get(id);
        if (p != null) {
            System.out.println("Product Found:\n" + p);
        } else {
            System.out.println("Product not found.");
        }
    }

    public void listProducts() {
        System.out.printf("%-10s %-20s %-10s %-10s%n", "ID", "Name", "Price", "Quantity");
        System.out.println("--------------------------------------------------------------");
        for (Product p : inventory.values()) {
            System.out.println(p);
        }
    }

    public void displayMenu() {
        System.out.println("\n--- Inventory Management ---");
        System.out.println("1. Add Product");
        System.out.println("2. Update Product");
        System.out.println("3. Delete Product");
        System.out.println("4. Search Product");
        System.out.println("5. List All Products");
        System.out.println("6. Save & Exit");
        System.out.print("Choose an option: ");
    }

    public void run() {
        loadFromFile();
        int choice;
        do {
            displayMenu();
            while (!scanner.hasNextInt()) {
                scanner.next(); // discard non-integer input
                System.out.print("Enter a valid option: ");
            }
            choice = scanner.nextInt();
            scanner.nextLine(); // consume newline

            switch (choice) {
                case 1 -> addProduct();
                case 2 -> updateProduct();
                case 3 -> deleteProduct();
                case 4 -> searchProduct();
                case 5 -> listProducts();
                case 6 -> {
                    saveToFile();
                    System.out.println("Data saved. Exiting...");
                }
                default -> System.out.println("Invalid option.");
            }
        } while (choice != 6);
    }

    public static void main(String[] args) {
        new InventoryManager().run();
    }
}
